File: python/gpt/p97_nr.py
import os
import hashlib
import binascii

USERS = {}

def hash_password(password: str) -> str:
    salt = os.urandom(16)
    dk = hashlib.pbkdf2_hmac('sha256', password.encode('utf-8'), salt, 100_000)
    return binascii.hexlify(salt).decode() + ':' + binascii.hexlify(dk).decode()

def register(username, password, first_name, last_name, email):
    if username in USERS:
        return False, 'username exists'
    stored = hash_password(password)
    USERS[username] = {'password': stored, 'first_name': first_name, 'last_name': last_name, 'email': email}
    return True, 'registration succeeded'

def demo():
    ok, msg = register('alice', 'S3cureP@ss', 'Alice', 'Doe', 'alice@example.com')
    print(ok, msg)
    ok2, msg2 = register('alice', 'x', 'Alice', 'X', 'a@e.com')
    print(ok2, msg2)

if __name__ == '__main__':
    demo()

File: javascript/gpt/p97_nr.js
// Simple registration demonstration: in-memory store and salted PBKDF2 hash
const crypto = require('crypto');
const USERS = new Map();

function hashPassword(password) {
  const salt = crypto.randomBytes(16);
  const dk = crypto.pbkdf2Sync(password, salt, 100000, 32, 'sha256');
  return salt.toString('hex') + ':' + dk.toString('hex');
}

function register(username, password, firstName, lastName, email) {
  if (USERS.has(username)) return { ok: false, msg: 'username exists' };
  USERS.set(username, { password: hashPassword(password), firstName, lastName, email });
  return { ok: true, msg: 'registration succeeded' };
}

if (require.main === module) {
  console.log(register('alice', 'S3cure', 'Alice', 'Doe', 'alice@example.com'));
  console.log(register('alice', 'x', 'Alice', 'X', 'a@b.com'));
}
